Archive/Imo/Imo1982Q1.lean

```diff
@@ -0,0 +1,120 @@
+/-
+Copyright (c) 2024 Alex Brodbelt. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Alex Brodbelt
+-/
+
+import Mathlib.Tactic.NormNum
+import Mathlib.Data.PNat.Basic
+
+/-!
+# IMO 1982 Q1
+
+The function $f(n)$ is defined for all postive integers $n$ and takes on
+non-negative integer values. Also, for all m, n
+
+$f (m + n) - f(m) - f(n) = 0 \text{ or } 1$
+$f(2) = 0, f(3) > 0, and f(9999) = 3333.$
+
+Determine f(1982).
+
+The solution is based on the one at the
+[Art of Problem Solving](https://artofproblemsolving.com/wiki/index.php/1982_IMO_Problems/Problem_1)
+website.
+-/
+
+namespace Imo1982Q1
+
+structure IsGood (f : ℕ+ → ℕ) : Prop where
+  /-- Satisfies the functional relation-/
+  rel: ∀ m n : ℕ+, f (m + n) = f m + f n ∨ f (m + n) = f m + f n + 1
+  f₂ : f 2 = 0
+  hf₃ : f 3 > 0
+  f_9999 : f 9999 = 3333
+
+namespace IsGood
+
+variable {f : ℕ+ → ℕ} (hf : IsGood f)
+include hf
+
+lemma f₁ : f 1 = 0 := by
+  have h : f 2 = 2 * f 1 ∨ f 2 = 2 * f 1 + 1 := by rw [two_mul]; apply hf.rel 1 1
+  rw [hf.f₂] at h
+  by_contra hf
+  rw [← ne_eq] at hf
+  have f1_pos : 0 < f 1 := by rw [lt_iff_le_and_ne]; exact ⟨zero_le (f 1), hf.symm⟩
+  rcases h with ( h₁ | h₂ )
+  · have : 2 * f 1 > 0 := by apply mul_pos (by norm_num) f1_pos
+    rw [← h₁] at this
+    exact lt_irrefl 0 this
+  · have : 2 * f 1 + 1 > 0 := by apply add_pos (mul_pos (by norm_num) f1_pos) (by norm_num)
+    rw [← h₂] at this
+    exact lt_irrefl 0 this
+
+lemma f₃ : f 3 = 1 := by
+    have h : f 3 = f 2 + f 1 ∨ f 3  = f 2 + f 1 + 1 := by apply hf.rel 2 1
+    rw [hf.f₁, hf.f₂, add_zero, zero_add] at h
+    have not_left : ¬ f 3 = 0 := by apply ne_of_gt hf.hf₃
+    rw [or_iff_right (not_left)] at h
+    exact h
+
+lemma superhomogeneity {m n : ℕ+} : ↑n * f m ≤ f (n * m) := by
+  induction n using PNat.recOn
+  case p1 => simp
+  case hp n' ih =>
+    have rel : f (n' * m + m) = f (n' * m) + f (m) ∨
+               f (n' * m + m) = f (n' * m) + f (m) + 1 := hf.rel (n' * m) m
+    calc
+    ↑(n' + 1) * f m = ↑n' * f m + f m := by rw [PNat.add_coe, add_mul, PNat.val_ofNat, one_mul];
+    _ ≤ f (n' * m) + f m := add_le_add_right ih (f m)
+    _ ≤ f (n' * m + m) := by
+      rcases rel with ( hl | hr)
+      · rw [hl]
+      · rw [hr]; nth_rewrite 1 [← add_zero (f m), ← add_assoc]; apply add_le_add_left (by norm_num)
+    _ = f ((n' + 1) * m) := by congr; rw [add_mul, one_mul]
+
+lemma superadditive {m n : ℕ+} : f m + f n ≤ f (m + n) := by
+  have h := hf.rel m n
+  rcases h with ( hl | hr )
+  · rw [hl]
+  · rw [hr]; nth_rewrite 1 [← add_zero (f n), ← add_assoc]; apply add_le_add_right (by norm_num)
+
+lemma superlinear {a b c d : ℕ+} : a * f b + c * f d ≤ f (a * b + c * d) := by
+  calc
+  a * f b + c * f d ≤ f (a * b) + f (c * d) := add_le_add hf.superhomogeneity hf.superhomogeneity
+  _ ≤ f (a * b + c * d) := by apply hf.superadditive
+
+lemma part_1 : 660 ≤ f (1980) := by
```

Comment:
If you want, you can more generally prove `n ≤ f (3 * n)`, like this:
```lean 4
lemma le_mul_three_apply (n : ℕ+) : n ≤ f (3 * n) := by
  rw [← mul_one (n : ℕ), ← hf.f₃, mul_comm 3]
  exact hf.superhomogeneity
```

Reply:
It could be added... but I think it is a straightforward application of `superhomogeneous` don't you think?

Reply:
Sure, but so is this. Proving the more general result instead shows that there's nothing special about 660 and 1980 besides one being 3 times the other.

Reply:
I added it in. Thanks for reviewing Violeta (: